linalg/elementmatrix.cpp

```diff
@@ -0,0 +1,286 @@
+      for (int i = 0; i < NDOFS; i++)
+      {
+         res += A(j, i, 1, f)*X(i, 1, f);
+      }
+      Y(j, 1, f) += res;
+   });
+}
+
+std::ostream& operator<<(std::ostream& os, const FaceMatrixInt& mat)
+{
+   for (size_t f = 0; f < mat.nf; f++)
+   {
+      os << "Face "<<f <<std::endl;
+      for (size_t i = 0; i < mat.ndofs; i++)
+      {
+         for (size_t j = 0; j < mat.ndofs; j++)
+         {
+            os << mat.data[i+j*mat.ndofs+2*f*mat.ndofs*mat.ndofs] << ", ";
+         }
+         os << std::endl;
+      }
+      os << std::endl;
+      for (size_t i = 0; i < mat.ndofs; i++)
+      {
+         for (size_t j = 0; j < mat.ndofs; j++)
+         {
+            os << mat.data[i+j*mat.ndofs+(2*f+1)*mat.ndofs*mat.ndofs] << ", ";
+         }
+         os << std::endl;
+      }
+      os << std::endl << std::endl;
+   }
+   return os;
+}
+
+void FaceMatrixExt::AddMult(const Vector &x, Vector &y) const
+{
+   auto X = Reshape(x.Read(), ndofs, 2, nf);
+   auto Y = Reshape(y.ReadWrite(), ndofs, 2, nf);
+   auto A = Reshape(data.Read(), ndofs, ndofs, 2, nf);
+   const int NDOFS = ndofs;
+   MFEM_FORALL(glob_j, nf*ndofs,
+   {
+      const int f = glob_j/NDOFS;
+      const int j = glob_j%NDOFS;
+      double res = 0.0;
+      for (int i = 0; i < NDOFS; i++)
+      {
+         res += A(i, j, 0, f)*X(i, 0, f);
+      }
+      Y(j, 1, f) += res;
+      res = 0.0;
+      for (int i = 0; i < NDOFS; i++)
+      {
+         res += A(i, j, 1, f)*X(i, 1, f);
+      }
+      Y(j, 0, f) += res;
+   });
+}
+
+void FaceMatrixExt::AddMultTranspose(const Vector &x, Vector &y) const
+{
+   auto X = Reshape(x.Read(), ndofs, 2, nf);
+   auto Y = Reshape(y.ReadWrite(), ndofs, 2, nf);
+   auto A = Reshape(data.Read(), ndofs, ndofs, 2, nf);
+   const int NDOFS = ndofs;
+   MFEM_FORALL(glob_j, nf*ndofs,
+   {
+      const int f = glob_j/NDOFS;
+      const int j = glob_j%NDOFS;
+      double res = 0.0;
+      for (int i = 0; i < NDOFS; i++)
+      {
+         res += A(j, i, 0, f)*X(i, 0, f);
+      }
+      Y(j, 1, f) += res;
+      res = 0.0;
+      for (int i = 0; i < NDOFS; i++)
+      {
+         res += A(j, i, 1, f)*X(i, 1, f);
+      }
+      Y(j, 0, f) += res;
+   });
+}
+
+std::ostream& operator<<(std::ostream& os, const FaceMatrixExt& mat)
+{
+   for (size_t f = 0; f < mat.nf; f++)
+   {
+      os << "Face "<<f <<std::endl;
+      for (size_t i = 0; i < mat.ndofs; i++)
+      {
+         for (size_t j = 0; j < mat.ndofs; j++)
+         {
+            os << mat.data[i+j*mat.ndofs+2*f*mat.ndofs*mat.ndofs] << ", ";
+         }
+         os << std::endl;
+      }
+      os << std::endl;
+      for (size_t i = 0; i < mat.ndofs; i++)
+      {
+         for (size_t j = 0; j < mat.ndofs; j++)
+         {
+            os << mat.data[i+j*mat.ndofs+(2*f+1)*mat.ndofs*mat.ndofs] << ", ";
+         }
+         os << std::endl;
+      }
+      os << std::endl << std::endl;
+   }
+   return os;
+}
+
+void FaceMatrixBdr::AddMult(const Vector &x, Vector &y) const
```

Comment:
I think this is exactly the same implementation as `ElementMatrix::AddMult` which means `ElementMatrix` and `FaceMatrixBdr` are actually the same class (you can remove this and just use ElementMatrix where you use it).